File: src/lotstretcher/imaging/compose/spin.py
"""
A rotating "spin" video from a vehicle's own real angle-labeled cutouts --
the used-inventory-appropriate alternative to a licensed generic 3D CAD
model (right for new inventory, wrong for used: it shows the SKU, not the
actual physical car with its actual wear).

WHAT THIS IS NOT: true 3D reconstruction (photogrammetry / NeRF / Gaussian
Splatting) of the real vehicle. That's confirmed research-stage industry-
wide as of 2025-2026 (see decision 1449fdef, lotstretcher project graph) -- even
the field's own state of the art needed a purpose-built moving-vehicle
capture rig, not reprocessing of static lot photos. Nobody ships that at
scale yet, lotstretcher included.

WHAT THIS IS: the same class of trick CarCutter's own documented "NextGen
360" actually uses -- synthesize a smooth rotation from a HANDFUL of real
photos rather than requiring a turntable shoot. CarCutter's version works
from 4 photos (front/rear/left/right) with a trained view-interpolation
model. This version is classical, not learned: cross-dissolve between the
real angle-labeled cutouts already sitting in a vehicle's cutout gallery
(lotstretcher typically has 5 distinct angles per vehicle -- front, front_3q,
side, rear_3q, rear -- MORE real anchor frames than CarCutter's own
baseline), scaled/anchored to a consistent canvas position so the car
doesn't visibly jump or resize between shots. It reads as a rotation, not
a slideshow, but it is honest about being an interpolation between real
photos rather than a synthesized new viewpoint -- there is no attempt to
paint in geometry the source photos don't show.

SCOPE: one side's half-turn (front through rear along whichever side has
the fuller angle coverage), not a full 360 -- completing the other side
needs real photos of it, which isn't guaranteed per vehicle. Extending to
a full loop (this side, then a hard cut/dissolve to the mirrored return)
is a reasonable next step once this baseline is validated against real
output.
"""
from __future__ import annotations

import subprocess
import sys
from collections.abc import Iterator
from pathlib import Path

from PIL import Image

from ... import core
from ... import spec as _spec

# From shared/pipeline-spec.json (the `spin` block), which core/src/spin.rs
# embeds: the same numbers drive both surfaces. The angle order is
# front-to-rear geometric, NOT the carousel's reveal order -- a spin has
# to read as continuous rotation, so it has to be monotonic.
SPIN_ANGLE_ORDER: list[str] = list(_spec.get("spin", "angleOrder"))
DEFAULT_CANVAS_SIZE = (_spec.get("spin", "canvasSize"), _spec.get("spin", "canvasSize"))
DEFAULT_FPS = float(_spec.get("spin", "fps"))
MIN_ANCHORS = int(_spec.get("spin", "minAnchors"))
DEFAULT_BUDGET_MB = float(_spec.get("spin", "budgetMb"))


def order_for_spin(cutout_dir: Path) -> list[tuple[Path, str]]:
    """Best (highest-confidence) real cutout per angle, in geometric
    front-to-rear order. Angles the vehicle doesn't have a shot for are
    just absent -- interpolating between whatever IS there is still
    honest; inventing a missing anchor would not be."""
    from ..select import load_angles

    cutout_dir = Path(cutout_dir)
    angles = load_angles(cutout_dir)
    if not angles:
        return []

    best: dict[str, tuple[str, float]] = {}
    for filename, info in angles.items():
        label, conf = info["angle"], info["confidence"]
        if label not in best or conf > best[label][1]:
            best[label] = (filename, conf)

    return [(cutout_dir / best[label][0], label) for label in SPIN_ANGLE_ORDER if label in best]


def plan_spin(cutouts: list[Image.Image], canvas_size: tuple[int, int], fps: float | None = None) -> dict:
    """The core's plan: one rect per anchor (one height, one ground line),
    the hold and dissolve lengths in frames, and the spotlight."""
    return core.call({"op": "spin_plan", "width": canvas_size[0], "height": canvas_size[1], "fps": fps,
                      "anchors": [{"width": c.width, "height": c.height} for c in cutouts]})


def render_spin_frames(cutouts: list[Image.Image], gradient_colors: tuple, canvas_size: tuple[int, int],
                       fps: float | None = None) -> Iterator[Image.Image]:
    """Every frame of the spin, RGB, in order. The core decides the
    schedule, places each anchor once into a canvas-sized layer, and
    dissolves between layers; this host only walks the frame count."""
    plan = plan_spin(cutouts, canvas_size, fps)
    w, h = canvas_size
    base_angle, start, end = gradient_colors
    # The layers and the backdrop stay resident in the core for the
    # whole clip: a frame is a request naming them by id, and the only
    # pixels that cross the boundary are the finished frame's. The
    # backdrop never turns in a spin, so it is rendered once.
    held: list[int] = []
    try:
        layers = []
        for c, rect in zip(cutouts, plan["rects"]):
            layer = core.retain(core.call({"op": "place_layer", "image": {"$image": 0}, "width": w, "height": h,
                                           "rect": rect}, [c.convert("RGBA")]))
            held.append(layer)
            layers.append(layer)
        backdrop = core.retain(core.render_frame([], w, h, {"kind": "linear", "angle": base_angle,
                                                            "start": list(start), "end": list(end)},
                                                 spotlight=tuple(plan["spotlight"])))
        held.append(backdrop)
        for f in range(plan["total_frames"]):
            at = core.call({"op": "spin_frame", "plan": plan, "frame": f})
            mix = None if at["next"] is None else (layers[at["next"]], at["tau"])
            yield core.render_frame([(layers[at["index"]], 0, 0, w, h, 1.0, mix)], w, h, {"kind": "image"},
                                    background_image=backdrop)
    finally:
        for image_id in held:
            core.release(image_id)


def render_spin_video(cutout_dir: Path, out_path: Path, gradient_colors: tuple,
                       canvas_size: tuple[int, int] = DEFAULT_CANVAS_SIZE,
                       fps: float = DEFAULT_FPS, budget_mb: float = DEFAULT_BUDGET_MB,
                       encoder: str = "libx264") -> dict | None:
    """Renders the spin and encodes it to `out_path`. Returns a small
    report dict, or None if there aren't enough real angles to make a
    spin worth producing (see MIN_ANCHORS) -- not an error, just not
    every vehicle's gallery has the coverage for this yet.

    Mirrors hero_video.py's render_hero_video()'s own encoder-fallback
    contract on purpose (retry once with libx264 if a hardware encoder
    fails to open under GPU pressure).
    """
    anchors = order_for_spin(cutout_dir)
    if len(anchors) < MIN_ANCHORS:
        return None

    cutouts = [Image.open(p).convert("RGBA") for p, _label in anchors]
    plan = plan_spin(cutouts, canvas_size, fps)
    total_frames, total_seconds = plan["total_frames"], plan["total_seconds"]
    n = len(cutouts)

    overhead = float(_spec.get("spin", "containerOverheadFrac"))
    bitrate_kbps = round((budget_mb * 8192) / total_seconds * overhead)

    out_path = Path(out_path)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    log_path = out_path.with_suffix(".ffmpeg.log")

    def _run_encode(enc: str) -> None:
        cmd = [
            "ffmpeg", "-y",
            "-f", "rawvideo", "-pix_fmt", "rgb24", "-s", f"{canvas_size[0]}x{canvas_size[1]}",
            "-r", str(fps), "-i", "pipe:0",
            "-c:v", enc, "-preset", "medium", "-b:v", f"{bitrate_kbps}k",
            "-maxrate", f"{int(bitrate_kbps * 1.2)}k", "-bufsize", f"{int(bitrate_kbps * 2)}k",
            "-pix_fmt", "yuv420p", "-t", f"{total_seconds:.3f}", str(out_path),
        ]
        with open(log_path, "w") as logf:
            proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.DEVNULL, stderr=logf)
            try:
                for frame in render_spin_frames(cutouts, gradient_colors, canvas_size, fps):
                    try:
                        proc.stdin.write(frame.tobytes())
                    except BrokenPipeError:
                        break
            finally:
                proc.stdin.close()
                ret = proc.wait()
        if ret != 0:
            raise RuntimeError(f"ffmpeg failed (exit {ret}) -- see {log_path}")

    enc_used = encoder
    try:
        _run_encode(encoder)
    except RuntimeError:
        if encoder == "libx264":
            raise
        print(f"    ! {encoder} failed for spin video, retrying with libx264", file=sys.stderr)
        enc_used = "libx264"
        _run_encode("libx264")

    log_path.unlink(missing_ok=True)
    return {
        "out_path": str(out_path),
        "encoder": enc_used,
        "duration_s": round(total_seconds, 2),
        "n_anchors": n,
        "frames": total_frames,
        "angles": [label for _p, label in anchors],
        "file_size_mb": round(out_path.stat().st_size / (1024 * 1024), 2),
    }

File: core/src/spin.rs
//! The spin choreography (port of the scheduling and placement half of
//! compose/spin.py): a half-turn cross-dissolved between a vehicle's
//! real angle cutouts. The host supplies the anchors in geometric order
//! (spec spin.angleOrder, chosen from angles.json), the core says where
//! each sits, how long it holds, and which two are on screen at any
//! frame. Pixels go through `place_layer`, `blend` and
//! frame::render_frame, so both surfaces produce the same clip.

use serde::{Deserialize, Serialize};

use crate::resize::resize_lanczos;
use crate::spec;
use crate::Image;

#[derive(Deserialize)]
pub struct AnchorSize { pub width: usize, pub height: usize }

#[derive(Deserialize)]
pub struct SpinPlanRequest {
    pub width: usize,
    pub height: usize,
    #[serde(default)]
    pub fps: Option<f64>,
    pub anchors: Vec<AnchorSize>,
}

#[derive(Serialize, Deserialize, Clone)]
pub struct SpinPlan {
    /// [x, y, w, h] each anchor is drawn at: one height, one ground line.
    pub rects: Vec<[i64; 4]>,
    pub fps: f64,
    pub hold_frames: usize,
    pub trans_frames: usize,
    pub total_frames: usize,
    pub total_seconds: f64,
    /// [cx, cy, dim] of the spotlight on the ground line.
    pub spotlight: [f64; 3],
}

#[derive(Serialize)]
pub struct SpinFrame {
    pub index: usize,
    /// The next anchor when the frame is inside a dissolve, else null.
    pub next: Option<usize>,
    pub tau: f64,
}

/// Python's round(): half to even, which is what the anchor sizes were
/// computed with.
fn round_even(v: f64) -> f64 {
    let r = v.round();
    if (v - v.trunc()).abs() == 0.5 && r % 2.0 != 0.0 { r - v.signum() } else { r }
}

pub fn plan(req: &SpinPlanRequest) -> Result<SpinPlan, String> {
    if req.width == 0 || req.height == 0 { return Err("canvas must not be empty".into()); }
    let (cw, ch) = (req.width as f64, req.height as f64);
    let fps = req.fps.unwrap_or_else(|| spec::f64_at(&["spin", "fps"]));
    let target_h = round_even(ch * spec::f64_at(&["spin", "heightFrac"]));
    let ground_y = round_even(ch * spec::f64_at(&["spin", "groundFrac"]));
    let mut rects = Vec::with_capacity(req.anchors.len());
    for a in &req.anchors {
        if a.height == 0 { return Err("anchor must not be empty".into()); }
        let scale = target_h / a.height as f64;
        let w = round_even(a.width as f64 * scale).max(1.0);
        let x = ((cw - w) / 2.0).floor();
        rects.push([x as i64, (ground_y - target_h) as i64, w as i64, target_h as i64]);
    }
    let hold_frames = round_even(spec::f64_at(&["spin", "holdSeconds"]) * fps) as usize;
    let trans_frames = round_even(spec::f64_at(&["spin", "transitionSeconds"]) * fps) as usize;
    let n = req.anchors.len();
    let total_frames = n * hold_frames + n.saturating_sub(1) * trans_frames;
    Ok(SpinPlan {
        rects, fps, hold_frames, trans_frames, total_frames,
        total_seconds: total_frames as f64 / fps,
        spotlight: [cw / 2.0, ch * spec::f64_at(&["spin", "groundFrac"]), spec::f64_at(&["spin", "spotlightDim"])],
    })
}

pub fn frame(plan: &SpinPlan, f: usize) -> SpinFrame {
    let n = plan.rects.len();
    let cycle = plan.hold_frames + plan.trans_frames;
    if n == 0 || cycle == 0 { return SpinFrame { index: 0, next: None, tau: 0.0 }; }
    let (idx, within) = (f / cycle, f % cycle);
    if idx >= n - 1 || within < plan.hold_frames {
        SpinFrame { index: idx.min(n - 1), next: None, tau: 0.0 }
    } else {
        SpinFrame { index: idx, next: Some(idx + 1), tau: (within - plan.hold_frames) as f64 / plan.trans_frames as f64 }
    }
}

/// A canvas-sized transparent RGBA layer with `car` resized to (w, h)
/// and copied in at (x, y): what the dissolve blends between.
pub fn place_layer(car: &Image, width: usize, height: usize, rect: [i64; 4]) -> Result<Image, String> {
    if car.channels != 4 { return Err("place_layer needs an RGBA car".into()); }
    let [x, y, w, h] = rect;
    let (w, h) = (w.max(1) as usize, h.max(1) as usize);
    let scaled = if w == car.width && h == car.height { car.clone() } else { resize_lanczos(car, w, h) };
    let mut out = Image::new(width, height, 4);
    for sy in 0..h {
        let dy = y + sy as i64;
        if dy < 0 || dy >= height as i64 { continue; }
        for sx in 0..w {
            let dx = x + sx as i64;
            if dx < 0 || dx >= width as i64 { continue; }
            let si = (sy * w + sx) * 4;
            let di = (dy as usize * width + dx as usize) * 4;
            out.data[di..di + 4].copy_from_slice(&scaled.data[si..si + 4]);
        }
    }
    Ok(out)
}

/// Pillow's Image.blend: a + (b - a) * t on every channel, alpha
/// included, in single precision and truncated to the pixel (measured
/// against Pillow 12: no half-up rounding).
pub fn blend(a: &Image, b: &Image, t: f64) -> Result<Image, String> {
    if a.width != b.width || a.height != b.height || a.channels != b.channels {
        return Err("blend needs two images of one shape".into());
    }
    let t = t as f32;
    let mut out = a.clone();
    if a.channels == 4 {
        // A layer is mostly untouched zeros: a pixel that is all zero in
        // both inputs blends to all zero, so only the rest is computed.
        for (o, (pa, pb)) in out.data.chunks_exact_mut(4).zip(a.data.chunks_exact(4).zip(b.data.chunks_exact(4))) {
            if pa == [0u8; 4] && pb == [0u8; 4] { continue; }
            for c in 0..4 {
                let (x, y) = (pa[c] as f32, pb[c] as f32);
                o[c] = (x + (y - x) * t).clamp(0.0, 255.0) as u8;
            }
        }
        return Ok(out);
    }
    for i in 0..out.data.len() {
        let (x, y) = (a.data[i] as f32, b.data[i] as f32);
        out.data[i] = (x + (y - x) * t).clamp(0.0, 255.0) as u8;
    }
    Ok(out)
}
